### PR TITLE
wxDC changer improvements

### PR DESCRIPTION
**Pull request by discnl**
This PR synchronizes the `wxDC` changer classes somewhat after seeing the `wxDC` attribute preservation changes of #2412, and does so with its first commit. The remaining commits could be more questionable.

The second commit ignores the (four times repeated) [doc note](https://github.com/wxWidgets/wxWidgets/blob/9028ae74fcdd020224ddcb3a3e98519608ca57b7/interface/wx/dc.h#L1986-L1987) regarding `Set()` only meant to be called once for `wxDC` changers. I couldn't think of a reason for such a restriction other than allowing to change `Set()` behaviour more freely later, but it already stores the attribute at most one time and by now ever changing that behaviour seems like a bad idea. Also wx itself already uses `Set()` twice in some minor instance (wxUniv code).

I slightly changed the docs but not for all `wxDC` changer classes yet as I was also hoping there's another solution besides maintaining docs for six classes that only differ by attribute.

The last commit is tentative: it collapses the six classes into a single macro-based one, which at least helps with synchronizing them. For me the culprit with making it template-based is `wxDCTextBgModeChanger` which is the only one not preserving a GDI object but an int instead.

Finally I also considered not adding `Set()` to `wxDCPenChanger` and `wxDCBrushChanger` and always preserving the attribute with each of the six changer classes but decided not to:

* While it would allow for code to keep on using `wxDC::SetXXX()` calls, by using a changer's `Set()` it is clearer from a single line that a DC changer is involved which I think is worth something readability-wise.

* There's enough instances (of `wxDCTextColourChanger`) where an attribute eventually doesn't get changed that I think it's wasteful to then always preserve an attribute. It would also undo the behaviour first documented in 489cc69b396e927bdd62da9970cfe64951071f92.

### CODE REVIEW

File: include/wx/dc.h [hunk 1]
```diff
@@ -1396,185 +1396,53 @@ class WXDLLIMPEXP_CORE wxDC : public wxObject
...
-private:
-    wxDC& m_dc;
-
-    wxColour m_colBgOld;
-
-    wxDECLARE_NO_COPY_CLASS(wxDCTextBgColourChanger);
-};
+// wxDCXXXChanger implementation
+
+#define wxDC_CHANGER(classname, attr_type, const_attr_type,  \
```
Comment (discnl) [inline include/wx/dc.h:1401]: The macro argument `const_attr_type` could be removed if using `const int&` instead of `int` in ctor and `Set()` of `wxDCTextBgModeChanger` is OK (I see no reason why not).

**Comment by vadz** [comment]
Thanks, I agree that the first commit makes sense, but I don't think using a macro is an improvement. Of course, we have plenty of those in wx code (and let's not look too closely at the commit history to see who wrote them...), but we really don't want to add any more. If we can use a common template for them without making things too complicated, let's do it, but if not leaving them as they';re now is still better than using macro IMO.

As for allowing `Set()` to be called twice, I don't like it much neither, but I could live with it if you think it's worth it. For me, these classes must be _simple_, so ideally they wouldn't have `Set()` at all. Unfortunately there are many cases where they need to be used conditionally, so we do need `Set()`. But the idea is still that you:

1. Create the "changer" object.
2. Initialize it with the new GDI object, either immediately in the ctor or very soon afterwards with `Set()`.
3. Don't do anything else with it until it's destroyed, restoring the original object.

IMHO it would be nice to keep them like this. With modern C++ we could impose the restriction of calling `Set()` only once at compile-time, I think, with rvalue-reference methods, but I don't think it can be done with C++98, unfortunately (but please let me know if you see some way to do it).

To summarize, the only thing I really object to here is putting all the code inside a macro because this makes it more difficult to read and to debug, but I'm also ambivalent about `Set()` changes.

**Comment by PBfordev** [comment]
BTW, is there a reason why there is no changer for the background brush (aside from "no one wrote it")? 

I am old so I am queasy about not restoring all DC objects I set but it seems that it now does not matter as it used to. But I have to admit that simplicity of GDI's `Save/RestoreDC()` had some appeal.

**Comment by discnl** [comment]
> Thanks, I agree that the first commit makes sense, but I don't think using a macro is an improvement. Of course, we have plenty of those in wx code (and let's not look too closely at the commit history to see who wrote them...), but we really don't want to add any more. If we can use a common template for them without making things too complicated, let's do it, but if not leaving them as they';re now is still better than using macro IMO.

Thanks for reviewing! I couldn't get it to work with a template, already getting stuck at the differently named `wxDC::SetXXX()` and `wxDC::GetXXX()` attribute functions. While it looks like that could be solved with overloading that seems to then require already having the changer classes which defeats the purpose somewhat.

> As for allowing `Set()` to be called twice, I don't like it much neither, but I could live with it if you think it's worth it. For me, these classes must be _simple_, so ideally they wouldn't have `Set()` at all. Unfortunately there are many cases where they need to be used conditionally, so we do need `Set()`.

The changers could do without `Set()` if they would always preserve the attribute, but as mentioned I don't think that's a good idea (for the conditional cases where an attribute perhaps doesn't change).
To preserve conditionally and without `Set()` could perhaps work when storing the old attribute at `wxDC::SetXXX()` time (if a `wxDC` was set to track that attribute by a changer). Not a particular pretty idea either, also because the functions are virtual, but then the changer classes would be more ideal without `Set()` ;).


> But the idea is still that you:
> 
>     1. Create the "changer" object.
> 
>     2. Initialize it with the new GDI object, either immediately in the ctor or very soon afterwards with `Set()`.
> 
>     3. Don't do anything else with it until it's destroyed, restoring the original object.

That's how I understood it to be used, but I didn't know of a good reason for 3 before. When drawing e.g. lines with different pens and no conditionals involved I then have to use:

```
wxDCPenChanger changer(dc, pen1)
dc.DrawLine()
dc.SetPen(pen2)
dc.DrawLine()
```

(or `Set()` and `SetPen()`, which to me is a less appealing option)

while for own code I would slightly prefer:

```
wxDCPenChanger changer(dc)
changer.Set(pen1)
dc.DrawLine()
changer.Set(pen2)
dc.DrawLine()
```

or otherwise:

```
wxDCPenChanger changer(dc, pen1)
dc.DrawLine()
changer.Set(pen2)
dc.DrawLine()
```

Neither of which is meant to be used currently.

If there would be no `Set()` then it could be:

```
wxDCPenChanger changer(dc)
dc.SetPen(pen1)
dc.DrawLine()
dc.SetPen(pen2)
dc.DrawLine()
```

Which may still be useful as `wxDCXXXRestorer` classes.

> To summarize, the only thing I really object to here is putting all the code inside a macro because this makes it more difficult to read and to debug, but I'm also ambivalent about `Set()` changes.

I don't like the macro either of course but disliked it slightly less than seeing the commits working towards repeating very similar code six times (and more similar again since the recent 75c473501622d61a0898a2e0647cfd5044938a1e). I'll remove the commit with the macro along with other changes later.

**Comment by discnl** [comment]
> BTW, is there a reason why there is no changer for the background brush (aside from "no one wrote it")?

Good one, looking at the getter and setters in dc.h that one may be missing and would complete the septet.

> I am old so I am queasy about not restoring all DC objects I set but it seems that it now does not matter as it used to. I have to admit that simplicity of GDI's `Save/RestoreDC()` had some appeal.

True, though I have no idea about its cost and availability of something similar on other platforms (and not just a common wx wrapper preserving each attribute). There's also a MSW private `wxPaintDCInfoExternal` making using of `Save/RestoreDC()` but it became unused since 025f5d1450a029e059807b1b0aba0233856f3891.
I don't know about the importance of restoring with Win32, IIRC not doing it doesn't result in resource leaks for e.g. pens but MSDN does still mention original objects should be selected again. I thought the (recent) preservation changes were done more to have the same `wxDC` state before and after calling a drawing function and not let callers make use of (or suffer from) implementation details, either on purpose or by accident.

**Comment by vadz** [comment]
I guess we could indeed have `wxDCPenRestorer` to allow code like the last snippet, but I admit I don't feel strongly enough about it to do anything myself.

**Comment by discnl** [comment]
> Thanks, I agree that the first commit makes sense, but I don't think using a macro is an improvement.

Are the template changes any better? At least it makes it easier to debug than the macro-based solution. The earliest I tested it with is MSVC9, it seems like using C++17 would allow for some shortcuts but that's of course not an option.

With this adding a `wxDCBgChanger` becomes trivial, and for `wxDCPenRestorer` (which so far is the only restorer that would have a use) the template needs duplicating with some changes.

**Comment by vadz** [comment]
I do think template-based version is better, although I'm still not quite convinced it's worth it just to avoid duplicating a few lines of pretty trivial code. But if you think it is, let's do it like this, however I'd like to get rid of the macro for defining the classes too and just write out the typedefs directly -- it's not much and it makes the code much more readable (for both humans and code navigation tools).

BTW why do we need the function pointer casts, are there any overloaded functions that need to be disambiguated?

**Comment by vadz** [comment]
This PR has been in "work needed" state for more than a year and will be closed soon. Please update it, e.g. by commenting here, if anybody intends to resume working on it.

**Comment by vadz** [comment]
I've included the first 2 commits in #25386, and will leave the other ones unapplied as their benefit is not really obvious.
